test/lib/jdk/test/lib/cds/CDSJarUtils.java

```diff
@@ -0,0 +1,358 @@
+            args.add("-C");
+            args.add(jarclassDir);
+            args.add(name + ".class");
+        }
+    }
+
+    /*
+     * This class is for passing extra options to the "jar" command-line tool
+     * for buildFromDirectory() and buildFromSourceDirectory().
+     *
+     * E.g.
+     *
+     * buildFromSourceDirectory("out.jar", "src", JarOptions.of().setMainClass("MyMainClass"),
+     *                          "--add-exports", "java.base/jdk.internal.misc=ALL-UNNAMED");
+     */
+    public static class JarOptions {
+        private String [] options;
+        private JarOptions() {}
+
+        public static JarOptions of(String... options) {
+            JarOptions jo = new JarOptions();
+            jo.options = options;
+            return jo;
+        }
+
+        public JarOptions setMainClass(String mainClass) {
+            if (mainClass != null) {
+                options = StringArrayUtils.concat(options, "--main-class=" + mainClass);
+            }
+            return this;
+        }
+        public JarOptions setManifest(String manifest) {
+            if (manifest != null) {
+                options = StringArrayUtils.concat(options, "--manifest=" + manifest);
+            }
+            return this;
+        }
+    }
+
+    public static void buildFromDirectory(String jarPath,
+                                          String classesDir) throws Exception {
+        buildFromDirectory(jarPath, classesDir, null);
+    }
+
+    public static void buildFromDirectory(String jarPath,
+                                          String classesDir,
+                                          JarOptions jarOptions) throws Exception {
+        ArrayList<String> argList = new ArrayList<String>();
+        argList.add("--create");
+        argList.add("--file=" + jarPath);
+        if (jarOptions != null) {
+            for (String s : jarOptions.options) {
+                argList.add(s);
+            }
+        }
+        argList.add("-C");
+        argList.add(classesDir);
+        argList.add(".");
+        createJar(argList);
+    }
+
+    public static void buildFromSourceDirectory(String jarName, String srcDir, String... extraJavacArgs) throws Exception  {
+        buildFromSourceDirectory(jarName, srcDir, null, extraJavacArgs);
+    }
+
+    /*
+     * Compile all source files under srcDir using javac with extraJavacArgs. Package
+     * all the classes into the specified JAR file
+     */
+    public static void buildFromSourceDirectory(String jarName, String srcDir, JarOptions jarOptions,
+                                                String... extraJavacArgs) throws Exception
+    {
+        System.out.print("Compiling " + srcDir + " into " + jarName);
+        if (extraJavacArgs.length > 0) {
+            System.out.print(" with");
+            for (String s : extraJavacArgs) {
+                System.out.print(" " + s);
+            }
+        }
+        System.out.println();
```

Comment:
Should this be printed only if DEBUG is enabled?

Reply:
I think it's better to always print the files being compiled, as jtreg also does the same thing for files that it compiles. That way it's easy to diagnose problems in the test source code.